src/main/java/wolox/training/controllers/BookController.java

```diff
@@ -122,23 +120,31 @@ public String greeting(
     @GetMapping
     @ResponseBody
     public ResponseEntity<Object> list(
-            @RequestParam(name = "isbn", required = false) Optional<String> isbn) {
-        if (isbn.isPresent()) {
-            return this.getBookByIsbn(isbn.get());
-        }
-        return new ResponseEntity<>(StreamSupport.stream(this.bookRepository.findAll().spliterator(), false)
-                .collect(Collectors.toList()), HttpStatus.OK);
-    }
+            @RequestParam(name = "isbn", required = false) String isbn,
+            @RequestParam(name = "genre", required = false) String genre,
+            @RequestParam(name = "author", required = false) String author,
+            @RequestParam(name = "image", required = false) String image,
+            @RequestParam(name = "title", required = false) String title,
+            @RequestParam(name = "subtitle", required = false) String subtitle,
+            @RequestParam(name = "publisher", required = false) String publisher,
+            @RequestParam(name = "year", required = false) String year,
+            @RequestParam(name = "pages", required = false) Integer pages) {
 
+        return new ResponseEntity<>(
+                this.bookRepository.findBooksBy(isbn, genre, author, image, title, subtitle, publisher, year, pages),
+                HttpStatus.OK);
+    }
 
-    private ResponseEntity<Object> getBookByIsbn(String isbn) {
-        Optional<Book> bookOptional = bookRepository.findFirstByIsbn(isbn);
+    @GetMapping("/isbn/{isbnNumber}")
+    private ResponseEntity<Object> getBookByIsbn(@PathVariable String  isbnNumber) {
```

Comment:
if we are sure that the PathVariable should be a number, we could set the var type as Integer instead of String (so the jackson parser throws exception when there is no number ). I think you should know about it, it helps a lot when you need to validate inputs ( and to avoid security issues ) 😁

Reply:
Perfect! And done :D The data type of isbn field is now long!